pkg/apis/controlplane/types.go

```diff
@@ -158,6 +159,9 @@ type NetworkPolicy struct {
 	TierPriority *int32
 	// Reference to the original NetworkPolicy that the internal NetworkPolicy is created for.
 	SourceRef *NetworkPolicyReference
+	// AppliedToPerRule tracks if appliedTo is set per rule basis rather than in policy spec.
+	// Must be false for K8s NetworkPolicy.
+	AppliedToPerRule bool
```

Comment:
I was wondering if you considered using `AppliedToGroups == nil` as the signal for this instead of introducing a new field, or if you don't think it is possible or a good idea?

Reply:
Sorry for the confusion here... You are misled by my poorly written testcase above. As you read from the actual implementation, the AppliedToGroups are set as **spec.appliedTo _or_ the union of all rule.appliedTo**, so it will never be nil. This is done to make the span calculation logic work as is. If we do not introduce the `AppliedToPerRule bool` field, then we need to introduce a separate field to track the union of all appliedTo in rules, which I personally think is harder to maintain compared to the current approach.

Reply:
I see. However, in the case where `AppliedToPerRule` is true, I believe the agent does not ever access the top-level `AppliedToGroups` field? If this is true, we would not really need to populate the field in the actual message sent from the controller to the agent (built by `ToNetworkPolicyMsg`). We only need the information in the object we store in the NetworkPolicy store in the controller. So in other words, I would image that both `AppliedToPerRule` and `AppliedToGroups` are needed in the type defined in `github.com/vmware-tanzu/antrea/pkg/controller/types` (storage) but not necessarily in the type defined in `github.com/vmware-tanzu/antrea/pkg/apis/controlplane` (message). By stripping the unneeded information from the message, we keep the size smaller. However, I may very well be mistaken here.

Reply:
The field is used in controlplane msg when the agent 'offloads' the controlplane rule. It uses this flag to properly set appliedTo for the agent internal rule data type. See cache.go, L532
```
func toRule(r *v1beta.NetworkPolicyRule, policy *v1beta.NetworkPolicy, maxPriority int32) *rule {
	appliedToGroups := policy.AppliedToGroups
	if policy.AppliedToPerRule {
		appliedToGroups = r.AppliedToGroups
	}
```
Before the change, we always do `appliedToGroups := policy.AppliedToGroups`. Now it might be the case that `r.AppliedToGroups` is a subset of `policy.AppliedToGroups`. But on a second thought, it seems that we just need to check if `policy.AppliedToGroups` is `nil` to make this decision... Will update the code.

Reply:
> But on a second thought, it seems that we just need to check if policy.AppliedToGroups is nil to make this decision... Will update the code.

yes that was my proposal. But it only works if the policy-level `AppliedToGroups` field is set to `nil` in `ToNetworkPolicyMsg` when the policy uses rule-level appliedTo

Reply:
Not necessarily. Just need to ensure that when appliedTo is set per policy scope, `AppliedToGroups` field in every rule is nil, and when appliedTo is set per rule, that field is not nil.